wikipron/extract/jpn.py

```diff
@@ -0,0 +1,229 @@
+    for pair_group in request.html.xpath(pair_check_path):
+        try:
+            word_element = pair_group.xpath(word_path)[0]
+        except IndexError:
+            return
+        word = word_element.text.rstrip(",(")
+        yield word
+
+
+# Japanese "Pronunciation" headers are most often sibling to a <ul> containing
+# an IPA transcription. There may also be a <ul> sibling to that <ul> which
+# contains a second IPA transcription (though second transcriptions are not
+# always contained within a separate <ul>). The only way I could work in
+# grabbing the transcriptions in both <ul>'s with my attempt to connect
+# words and prons "locally" - by moving stepwise from the word to the pron -
+# was to update _PRON_XPATH_SELECTOR with the second <ul> and
+# run a separate request.
+def _yield_jpn_upper_pron(
+    request: requests.Response, config: "Config", word_target: str
+) -> "Iterator[Pron]":
+    pron_path = _PRON_XPATH_SELECTOR.format(
+        word_to_work_from=word_target,
+        heading=_check_etymologies(request),
+        second_ul="[preceding-sibling::*[1][self::ul]]",
+    )
+    try:
+        request.html.xpath(pron_path)[0]
+    except IndexError:
+        return
+
+    for upper_pron_ele in request.html.xpath(pron_path):
+        prons = []
+        for pron in yield_pron(upper_pron_ele, IPA_XPATH_SELECTOR, config):
+            prons.append(pron)
+        yield prons
+
+
+def _yield_jpn_lower_pron(
+    request: requests.Response, config: "Config", word_target: str
+) -> "Iterator[Pron]":
+    pron_path = _PRON_XPATH_SELECTOR.format(
+        word_to_work_from=word_target,
+        heading=_check_etymologies(request),
+        second_ul="",
+    )
+    for pron_element in request.html.xpath(pron_path):
+        prons = []
+        for pron in yield_pron(pron_element, IPA_XPATH_SELECTOR, config):
+            prons.append(pron)
+        upper_prons = _yield_jpn_upper_pron(request, config, word_target)
+        try:
+            # Possible, though seemingly unlikely, bug here. This could
+            # potentially match a second "upper" pronunciation (most likely
+            # in a second Etymology) and append it to the "lower"
+            # pronunciations yielded from the first Etymology. Or append
+            # the "upper" pronunciation of the first Etymology to the
+            # "lower" pronunciation of the second Etymology, etc.
+            # Fortunately entries with multiple <ul>'s and multiple
+            # etymologies are exceedingly rare. I'm not sure I've seen
+            # any.
+            prons += next(upper_prons)
+        except StopIteration:
+            pass
+        # Yielding here because we don't want to collect all pronunciation
+        # entries on a page at the same time. Doing so would make it difficult
+        # to connect words to their prons appropriately. We only want to grab
+        # the prons that are linked to the word we have scraped.
+        yield prons
```

Comment:
We probably don't have to catch `StopIteration`, if we use `itertools.chain`:

```python
for pron_element in request.html.xpath(pron_path):
    lower_prons = yield_pron(...)
    upper_prons = _yield_jpn_upper_pron(...)
    # The following won't throw an error even if either or both of lower_prons and upper_prons are empty
    prons = itertools.chain(lower_prons, upper_prons)
    yield from prons
```

Reply:
I like the `itertools.chain()` approach, but I'm hesitant to exhaustively run `_yield_jpn_upper_pron(...)` each iteration of the loop. Then it would collect all the upper prons from all the Etymologies each time we grab the lower pron or prons for a specific word, potentially leading to a mismatch of words and pronunciations. (I say potentially because I've never seen any page with multiple etymologies that have multiple <ul>'s.)